I6358 support aria current

feerrenrut opened the pull request:
Support for aria current. Used http://ljwatson.github.io/design-patterns/aria-current/ to test this, as described we report just 'current' if `True` and 'current step' etc if there is another value.

Fixes #6358

feerrenrut committed:
Support for aria-current in focus mode for chrome

feerrenrut committed:
Support for aria-current in focus mode for IE11

feerrenrut committed:
Add support for aria-current in focus mode to Edge

feerrenrut committed:
Added support for aria-current values with edge

While in focus mode, we can now get the values for aria-current rather
than just the presence of the property.

feerrenrut committed:
aria current working in IE while in browse mode

feerrenrut committed:
Aria-current now reported in browse mode for Edge

feerrenrut committed:
Support other values of aria-current

feerrenrut commented:
aria-current is not yet being reported for list items on any browser.

@jcsteh mentioned yesterday that:

> List items are probably treated as layout by TextInfo.getPresentationCategory. A bit tricky because we don't want other states like read-only or the role to get reported, only really aria-current

jcsteh commented on source/speech.py:
```diff
@@ -1112,7 +1129,7 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
+		out.extend(x for x in (nameText,(stateText if speakStatesFirst else roleText),(roleText if speakStatesFirst else stateText),ariaCurrentText,valueText,descriptionText,levelText,keyboardShortcutText) if x)
 		if content and not speakContentFirst:
 			out.append(content)
 		return CHUNK_SEPARATOR.join(out)
```
Regarding speaking for list items, etc., this can be achieved by handling `current` similar to the way we handle `clickable` below. The conditions for current should be the same as clickable except we do want to report it even if extraDetail is True (whereas we don't for clickable). The tricky thing here is that ideally, we still want to report current if we're reporting clickable (and vice versa). So, I guess the clickable and current block will have to be merged.

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -257,6 +257,12 @@ def _getParagraphOffsets(self,offset):
 		return lineStart.value,lineEnd.value
 
 	def _normalizeControlField(self,attrs):
+
```
Extraneous blank line.

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -257,6 +257,12 @@ def _getParagraphOffsets(self,offset):
 
 	def _normalizeControlField(self,attrs):
+
+		ariaCurrent = attrs.get("IAccessible2::attribute_current")
```
Please move this into the gecko_ia2 virtual buffer.

jcsteh commented on source/virtualBuffers/__init__.py:
```diff
@@ -257,6 +257,12 @@ def _getParagraphOffsets(self,offset):
+		ariaCurrent = attrs.get("IAccessible2::attribute_current")
+		if ariaCurrent != None:
+			attrs['current']= ariaCurrent
+			del attrs["IAccessible2::attribute_current"]
```
Probably no need to delete this; we don't do it with other attributes and leaving it there might be useful for debugging.

jcsteh commented on source/speech.py:
```diff
@@ -1112,7 +1129,7 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
 		if content and speakContentFirst:
 			out.append(content)
-		out.extend(x for x in (nameText,(stateText if speakStatesFirst else roleText),(roleText if speakStatesFirst else stateText),valueText,descriptionText,levelText,keyboardShortcutText) if x)
+		out.extend(x for x in (nameText,(stateText if speakStatesFirst else roleText),(roleText if speakStatesFirst else stateText),ariaCurrentText,valueText,descriptionText,levelText,keyboardShortcutText) if x)
```
I think it's possible that someone might set aria-current on a table cell; e.g. in a calendar. So, it also needs to be reported for the table cell case above this.

jcsteh reviewed (changes requested):
Nice work so far. Thanks. :)

Aside from the review comments, support also needs to be implemented for braille. You'll want to look at braille.getBrailleTextForProperties, braille.NVDAObjectRegion.update and braille.getControlFieldBraille. The changes will be similar to those in speech. I realise you don't have a braille display, but you can use the [Braille Viewer driver](https://dl.dropboxusercontent.com/u/28976681/brailleViewer.py?dl=1) I hacked together some time ago to get braille enabled and then look at the IO log output to see the text being translated.

feerrenrut commented on source/NVDAObjects/__init__.py:
```diff
@@ -798,6 +798,12 @@ def _get_statusBar(self):
+	def getValueForAriaCurrent(self):
+		"""Gets the value for aria-current. Normally returns False. If this object is current
+		it will return one of the following values: True, "page", "step", "location", "date", "time"
+		"""
```
Yeah ok. I'm not 100% comfortable with the name, but I can't think of anything better.

feerrenrut committed:
Review actions for #6860

- ensure that aria-current is output to braille.
- make `getValueForAriaCurrent` a property called isCurrent
- use a compiled regex
- simplify `if` statement
- move text labels for 'aria-current' values to a dictionary and
  translate
- handle aria-current in list items and tables

feerrenrut commented:
To test this (with clickable list items, and clickable table cells) I used the following HTML file (unzip first)
[aria-current-test.zip](https://github.com/nvaccess/nvda/files/793794/aria-current-test.zip)

jcsteh commented on source/braille.py:
```diff
@@ -630,9 +633,16 @@ def getBrailleTextForProperties(**propertyValues):
 			# Translators: Displayed in braille for a table cell column number.
 			# %s is replaced with the column number.
 			textList.append(_("c%s") % columnNumber)
+	ariaCurrent = propertyValues.get('current', False)
```
nit: Please rename to current.

jcsteh commented on source/braille.py:
```diff
@@ -698,12 +708,16 @@ def getControlFieldBraille(info, field, ancestors, reportStart, formatConfig):
 	role = field.get("role", controlTypes.ROLE_UNKNOWN)
 	states = field.get("states", set())
 	value=field.get('value',None)
+	ariaCurrent=field.get('current', None)
```
Please rename.

jcsteh commented on source/controlTypes.py:
```diff
@@ -613,6 +613,23 @@
 #}
 
+# Text to use for 'current' values. These describe if an item is the current item 
+# within a particular kind of selection.
```
Please change the comment prefix to `#:`:

```
#: l1
#: l2
```

This means it'll be treated as the documentation of the variable by epydoc (the code doc tool).

jcsteh reviewed (approved):
Looks good! Ta!

feerrenrut committed:
Further review actions for #6860
